### DISCUSSION TITLE
Masking trajectories of different lengths

### DISCUSSION THREAD

**Discussion by joshuaspear** [thomaspinder/GPJax#770] [Q&A]
What is the most "gpjax" way to handle trajectories of different lengths? I have a dataset of one-dimensional time indexed trajectories. I would like to experiment with multi-task kernels i.e., defining a GP prior per trajectory and then specifying a cross-trajectory kernel. However, I am unsure how to implement this, taking into account the differing trajectory lengths?

**Comment by joshuaspear** [comment]
Hi - thank you so much for your reply! I went for the second method and it worked great! Josh Sent from my iPhoneOn 11 Aug 2026, at 14:58, Mohammad Rezwan Khan ***@***.***> wrote:﻿
The important constraint here is that GPJax's current multi-output API is rectangular: it expects the same number of input rows for every output. Dataset requires 2-D X and y with the same first dimension, and MultiOutputGaussian takes one num_datapoints=N and reshapes y.T into output-major [y_1(t_1..t_N), ..., y_P(t_1..t_N)]. The ICM example therefore uses one shared x grid with y.shape == (N, P).
So I would handle it in one of these two ways:

If the trajectories can be evaluated/interpolated on a common time grid, choose the grid and build a rectangular dataset:

# t has shape (N,); each column is one trajectory on that grid
X = t[:, None]
Y = jnp.column_stack([trajectory_on_grid[i] for i in range(P)])  # (N, P)
D = gpx.Dataset(X=X, y=Y)

coreg = gpx.parameters.CoregionalizationMatrix(num_outputs=P, rank=1, key=key)
kernel = gpx.kernels.ICMKernel(
    base_kernel=gpx.kernels.RBF(),
    coregionalization_matrix=coreg,
)
likelihood = gpx.likelihoods.MultiOutputGaussian(
    num_datapoints=N, num_outputs=P
)
Do not pad the shorter trajectories with NaN values: there is no mask field in Dataset, so those values would enter the marginal likelihood. Also avoid treating extrapolated tails as exact observations; restrict the grid to the shared support or model interpolation uncertainty separately.

If the observations are genuinely irregular/ragged and you need to retain every observed point, use a long-form model instead of the current ICMKernel/MultiOutputGaussian layout. Keep rows such as (time, trajectory_id, value) and use a covariance of the form

K((t, i), (t', j)) = B[i, j] * k_time(t, t')

with y shaped (M, 1). That requires a custom kernel/likelihood (or an extension to GPJax's multi-output compute path) that accepts discrete task IDs and missing output/task pairs. The current ICMKernel does not interpret an extra trajectory_id column as a discrete task; putting it into an ordinary RBF would instead make task IDs a continuous input distance. If cross-trajectory correlation is not needed, fitting one ordinary Dataset per trajectory is the simpler option.
The relevant implementation details are in Dataset, MultiOutputGaussian, the multi-output example, and ICMKernel. The example's predictions are output-major as well, so reshape a flat prediction with pred.mean.reshape(P, M).T.
